File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/camera/CameraOptions.java
package org.maplibre.nativeffi.camera;

import java.util.Objects;
import org.maplibre.nativeffi.geo.LatLng;
import org.maplibre.nativeffi.geo.ScreenPoint;

/** Mutable camera descriptor used for camera snapshots and commands. */
public final class CameraOptions {
  private LatLng center;
  private Double centerAltitude;
  private EdgeInsets padding;
  private ScreenPoint anchor;
  private Double zoom;
  private Double bearing;
  private Double pitch;
  private Double roll;
  private Double fieldOfView;

  public boolean hasCenter() {
    return center != null;
  }

  public LatLng center() {
    return center;
  }

  public CameraOptions setCenter(LatLng center) {
    this.center = Objects.requireNonNull(center, "center");
    return this;
  }

  public CameraOptions setCenter(double latitude, double longitude) {
    return setCenter(new LatLng(latitude, longitude));
  }

  public CameraOptions clearCenter() {
    center = null;
    return this;
  }

  public boolean hasCenterAltitude() {
    return centerAltitude != null;
  }

  public Double centerAltitude() {
    return centerAltitude;
  }

  public CameraOptions setCenterAltitude(double centerAltitude) {
    this.centerAltitude = centerAltitude;
    return this;
  }

  public CameraOptions clearCenterAltitude() {
    centerAltitude = null;
    return this;
  }

  public boolean hasPadding() {
    return padding != null;
  }

  public EdgeInsets padding() {
    return padding;
  }

  public CameraOptions setPadding(EdgeInsets padding) {
    this.padding = Objects.requireNonNull(padding, "padding");
    return this;
  }

  public CameraOptions clearPadding() {
    padding = null;
    return this;
  }

  public boolean hasAnchor() {
    return anchor != null;
  }

  public ScreenPoint anchor() {
    return anchor;
  }

  public CameraOptions setAnchor(ScreenPoint anchor) {
    this.anchor = Objects.requireNonNull(anchor, "anchor");
    return this;
  }

  public CameraOptions clearAnchor() {
    anchor = null;
    return this;
  }

  public boolean hasZoom() {
    return zoom != null;
  }

  public Double zoom() {
    return zoom;
  }

  public CameraOptions setZoom(double zoom) {
    this.zoom = zoom;
    return this;
  }

  public CameraOptions clearZoom() {
    zoom = null;
    return this;
  }

  public boolean hasBearing() {
    return bearing != null;
  }

  public Double bearing() {
    return bearing;
  }

  public CameraOptions setBearing(double bearing) {
    this.bearing = bearing;
    return this;
  }

  public CameraOptions clearBearing() {
    bearing = null;
    return this;
  }

  public boolean hasPitch() {
    return pitch != null;
  }

  public Double pitch() {
    return pitch;
  }

  public CameraOptions setPitch(double pitch) {
    this.pitch = pitch;
    return this;
  }

  public CameraOptions clearPitch() {
    pitch = null;
    return this;
  }

  public boolean hasRoll() {
    return roll != null;
  }

  public Double roll() {
    return roll;
  }

  public CameraOptions setRoll(double roll) {
    this.roll = roll;
    return this;
  }

  public CameraOptions clearRoll() {
    roll = null;
    return this;
  }

  public boolean hasFieldOfView() {
    return fieldOfView != null;
  }

  public Double fieldOfView() {
    return fieldOfView;
  }

  public CameraOptions setFieldOfView(double fieldOfView) {
    this.fieldOfView = fieldOfView;
    return this;
  }

  public CameraOptions clearFieldOfView() {
    fieldOfView = null;
    return this;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/camera/CameraOptions.kt
package org.maplibre.nativeffi.camera

import org.maplibre.nativeffi.geo.LatLng
import org.maplibre.nativeffi.geo.ScreenPoint

/**
 * Mutable camera descriptor used for camera snapshots and commands.
 *
 * Compares and hashes by field value; [copy] returns an independent instance. Keep an instance
 * unmodified while it is a key in a hash-based collection.
 */
public class CameraOptions {
  public var center: LatLng? = null

  public var centerAltitude: Double? = null

  public var padding: EdgeInsets? = null

  /**
   * The screen point that stays fixed while a camera command applies. Input-only: snapshots omit
   * this field.
   */
  public var anchor: ScreenPoint? = null

  public var zoom: Double? = null

  public var bearing: Double? = null

  public var pitch: Double? = null

  public var roll: Double? = null

  public var fieldOfView: Double? = null

  /** Returns an independent copy of this descriptor with [block] applied to the copy. */
  public fun copy(block: CameraOptions.() -> Unit = {}): CameraOptions =
    CameraOptions()
      .also {
        it.center = center
        it.centerAltitude = centerAltitude
        it.padding = padding
        it.anchor = anchor
        it.zoom = zoom
        it.bearing = bearing
        it.pitch = pitch
        it.roll = roll
        it.fieldOfView = fieldOfView
      }
      .apply(block)

  private val fields: List<Any?>
    get() = listOf(center, centerAltitude, padding, anchor, zoom, bearing, pitch, roll, fieldOfView)

  override fun equals(other: Any?): Boolean = other is CameraOptions && fields == other.fields

  override fun hashCode(): Int = fields.hashCode()
}
